### PR TITLE
Introduce @SearchExtension to configure Hibernate Search through annotated beans

### LINKED ISSUE #24922: @SearchExtension: equivalent to @PersistenceUnitExtension for Hibernate Search
### Description

We have a `@PersistenceUnitExtension` annotation to assign user-defined beans to the configuration of a persistence unit in Hibernate ORM.

We should add a similar annotation, say `@SearchExtension`, for Hibernate Search configuration.

Examples:

```
@SearchExtension
public class MyAnalysisConfigurer extends ElasticsearchAnalysisConfigurer {
    // ...
}
```

```
@SearchExtension(backend = "someNamedBackend")
public class MyAnalysisConfigurer extends ElasticsearchAnalysisConfigurer {
    // ...
}
```

Same for other configuration properties accepting beans:

* [quarkus.hibernate-search-orm.background-failure-handler](https://quarkus.io/guides/hibernate-search-orm-elasticsearch#quarkus-hibernate-search-orm-elasticsearch_quarkus.hibernate-search-orm.background-failure-handler)
* [quarkus.hibernate-search-orm.elasticsearch.layout.strategy](https://quarkus.io/guides/hibernate-search-orm-elasticsearch#quarkus-hibernate-search-orm-elasticsearch_quarkus.hibernate-search-orm.elasticsearch.layout.strategy)
* [quarkus.hibernate-search-orm.automatic-indexing.synchronization.strategy](https://quarkus.io/guides/hibernate-search-orm-elasticsearch#quarkus-hibernate-search-orm-elasticsearch_quarkus.hibernate-search-orm.automatic-indexing.synchronization.strategy) (Maybe?)

### Implementation ideas

_No response_

### PR DESCRIPTION
**Pull request by yrodiere**
Fixes #24922 

~~Creating this PR as draft, because it is based on #26186 , which must be merged first.~~ => Done

### CODE REVIEW

File: core/deployment/src/main/java/io/quarkus/deployment/steps/DevServicesConfigBuildStep.java [hunk 1]
```diff
@@ -60,10 +62,21 @@ public void run() {
...
         }
-        for (Map.Entry<String, String> entry : newProperties.entrySet()) {
+        oldConfig = newProperties;
+
+        Map<String, String> newPropertiesWithAdditionalConfig = new HashMap<>(newProperties);
+        // On contrary to dev services config, "additional" config build items are
+        // produced on each restart, so we don't want to remember them from one restart to the next.
+        for (DevServicesAdditionalConfigBuildItem item : devServicesAdditionalConfigBuildItems) {
+            if (newProperties.containsKey(item.getTriggeringKey())) {
+                item.getCallbackWhenEnabled().run();
```
Comment (gastaldi) [inline core/deployment/src/main/java/io/quarkus/deployment/steps/DevServicesConfigBuildStep.java:72]: Potential NPE here

Comment (yrodiere) [reply core/deployment/src/main/java/io/quarkus/deployment/steps/DevServicesConfigBuildStep.java:72]: Not at the moment, but I guess you never know where people will try to stuff a `null`... I'll add a check.

**Comment by yrodiere** [comment]
I've seen the failure and am currently working on solving it.

**Comment by yrodiere** [comment]
Hey @Sanne @gsmet , could one of you please review this PR? Thanks in advance :)

**Review by gsmet — APPROVED** [review]

**Pull request merged**